### PR TITLE
feat(insured-bridge-bot): instant relay logic

### PR DESCRIPTION
**Pull request by chrismaree**
**Motivation**

The bots need the ability to instantly relay deposits. This PR adds this functionality.

**Summary**

Instant relay is done via multicall. this enables the bots to call: a) `relayDeposit` and b) `speedUpRelay` within one transaction. To facilitate this, the bridge pool was modified to be multicallable.

This PR also made a small refactor to how transactions are generated within the relayer to be more reusable and to not repeat any code.

**Testing**

Check a box to describe how you tested these changes and list the steps for reviewers to test.

- [ ]  Ran end-to-end test, running the code as in production
- [X]  New unit tests created
- [ ]  Existing tests adequate, no new tests required
- [ ]  All existing tests pass
- [ ]  Untested

### CODE REVIEW

File: packages/insured-bridge-relayer/src/Relayer.ts [hunk 1]
```diff
@@ -150,25 +150,12 @@ export class Relayer {
...
-      realizedLpFee
-    );
-
+  private async slowRelay(deposit: Deposit, realizedLpFeePct: BN) {
     await this.gasEstimator.update();
     try {
       const { receipt, transactionConfig } = await runTransaction({
         web3: this.l1Client.l1Web3,
-        transaction: (slowRelayTx as unknown) as TransactionType,
+        transaction: this.generateSlowRelayTx(deposit, realizedLpFeePct),
```
Comment (chrismaree) [inline packages/insured-bridge-relayer/src/Relayer.ts:158]: tx generation done in a method below now.

File: packages/insured-bridge-relayer/src/Relayer.ts [hunk 2]
```diff
@@ -228,8 +212,59 @@ export class Relayer {
...
+      realizedLpFeePct
+    ) as unknown) as TransactionType;
+  }
+
+  private generateSpeedUpRelayTx(deposit: Deposit): TransactionType {
+    const bridgePool = this.l1Client.getBridgePoolForDeposit(deposit);
+    return (bridgePool.methods.speedUpRelay(deposit as any) as unknown) as TransactionType;
+  }
+
+  private generateInstantRelayTx(deposit: Deposit, realizedLpFeePct: BN): TransactionType {
```
Comment (chrismaree) [inline packages/insured-bridge-relayer/src/Relayer.ts:259]: the magic of multicall. we can re-use the previous methods as well as instant relay is simply the combination of a slow and speed up relay :)

File: packages/insured-bridge-relayer/src/Relayer.ts [hunk 3]
```diff
@@ -228,8 +212,59 @@ export class Relayer {
...
+      deposit.slowRelayFeePct,
+      deposit.instantRelayFeePct,
+      deposit.quoteTimestamp,
+      realizedLpFeePct
+    ) as unknown) as TransactionType;
+  }
+
+  private generateSpeedUpRelayTx(deposit: Deposit): TransactionType {
+    const bridgePool = this.l1Client.getBridgePoolForDeposit(deposit);
+    return (bridgePool.methods.speedUpRelay(deposit as any) as unknown) as TransactionType;
```
Comment (mrice32) [inline packages/insured-bridge-relayer/src/Relayer.ts:256]: Nit: Is the any cast necessary? Unknown should be sufficient, no?

File: packages/insured-bridge-relayer/src/Relayer.ts [hunk 4]
```diff
@@ -238,7 +273,9 @@ export class Relayer {
     realizedLpFeePct: BN
   ): { slow: BN; instant: BN } {
     return {
+      // slow relay: proposer bond = amount * proposerBondPct
```
Comment (mrice32) [inline packages/insured-bridge-relayer/src/Relayer.ts:276]: +1

File: packages/core/contracts/common/implementation/MultiCaller.sol [hunk 5]
```diff
@@ -0,0 +1,27 @@
+// This contract is taken from Uniswaps's multi call implementation (https://github.com/Uniswap/uniswap-v3-periphery/blob/main/contracts/base/Multicall.sol)
+// and was modified to be solidity 0.8 compatible. No other changes were made.
+pragma solidity ^0.8.0;
+
+// pragma abicoder v2;
```
Comment (mrice32) [inline packages/core/contracts/common/implementation/MultiCaller.sol:5]: Nit: comment?

**Review by mrice32 — APPROVED** [review]
LGTM!

File: packages/insured-bridge-relayer/src/Relayer.ts (hunk 3 above)
Comment (chrismaree) [reply packages/insured-bridge-relayer/src/Relayer.ts:256]: without that additional type cast I get this:
image

**Pull request merged**